lib/policies/exponentialRetryPolicy.ts

```diff
@@ -0,0 +1,126 @@
+import { HttpResponse } from "../httpResponse";
+import { BaseRequestPolicy, RequestPolicy } from "../requestPolicy";
+import { RequestPolicyFactory } from "../requestPolicyFactory";
+import { RequestPolicyOptions } from "../requestPolicyOptions";
+import * as utils from "../util/utils";
+
+/**
+ * An error that can be thrown when the maximum number of attempts have been attempted.
+ */
+export interface RetryError extends Error {
+    /**
+     * The message that describes the retry failure.
+     */
+    message: string;
+    /**
+     * The error code that describes the retry failure.
+     */
+    code?: string;
+    /**
+     * An inner error that this RetryError wraps.
+     */
+    innerError?: RetryError;
+}
+
+/**
+ * The options that can be passed to an ExponentialRetryPolicy.
+ */
+export interface RetryOptions {
+    /**
+     * The maximum number of attempts that the retry policy will perform before failing. The first
+     * attempt after a failure is considered the 2nd attempt.
+     */
+    maximumAttempts?: number;
+    /**
+     * The number of milliseconds to delay before attempting again.
+     */
+    initialRetryDelayInMilliseconds?: number;
+    /**
+     * The maximum number of milliseconds to wait before retrying.
+     */
+    maximumRetryIntervalInMilliseconds?: number;
+    /**
+     * The function to use to delay before sending a retry attempt.
+     */
+    delayFunction?: (delayInMilliseconds: number) => Promise<void>;
+}
+
+/**
+ * Get a RequestPolicyFactory that creates ExponentialRetryPolicies.
+ */
+export function exponentialRetryPolicy(retryOptions?: RetryOptions): RequestPolicyFactory {
+    return (nextPolicy: RequestPolicy, options: RequestPolicyOptions) => {
+        return new ExponentialRetryPolicy(retryOptions || {}, nextPolicy, options);
+    };
+}
+
+export class ExponentialRetryPolicy extends BaseRequestPolicy {
+    private readonly _maximumAttempts: number;
+    private readonly _initialRetryDelayInMilliseconds: number;
+    private readonly _maximumRetryDelayInMilliseconds: number;
+    private readonly _delayFunction: (delayInMilliseconds: number) => Promise<void>;
+
+    constructor(retryOptions: RetryOptions, nextPolicy: RequestPolicy, options: RequestPolicyOptions) {
+        super(nextPolicy, options);
+
+        this._maximumAttempts = retryOptions.maximumAttempts || 3;
+        this._initialRetryDelayInMilliseconds = retryOptions.initialRetryDelayInMilliseconds || 30 * 1000;
+        this._maximumRetryDelayInMilliseconds = retryOptions.maximumRetryIntervalInMilliseconds || 90 * 1000;
+        this._delayFunction = retryOptions.delayFunction || utils.delay;
+    }
+
+    /**
+     * Get whether or not we should retry the request based on the provided response.
+     * @param response The response to read to determine whether or not we should retry.
+     */
+    protected shouldRetry(details: { response?: HttpResponse, responseError?: RetryError}): boolean {
+        let result = true;
```

Comment:
General comment: make sure the indentation is 2 spaces. That is the norm in JS/TS world.

Reply:
Done.